packages/styled-components/src/models/StyledComponent.ts

```diff
@@ -29,24 +31,40 @@ import isTag from '../utils/isTag';
 import joinStrings from '../utils/joinStrings';
 import merge from '../utils/mixinDeep';
 import ComponentStyle from './ComponentStyle';
+import Keyframes from './Keyframes';
 import { useStyleSheet, useStylis } from './StyleSheetManager';
 import { ThemeContext } from './ThemeProvider';
 
-const identifiers: { [key: string]: number } = {};
+/** Shared context for component identifiers, preventing duplicate ids */
+const COMPONENT_IDENTIFIERS: Record<string, Record<string, number>> = {}
```

Comment:
The new identifier collection shape is like this:

```js
{
  'sc': {
    'hash': 0
  }
}
```

where `'sc'` is the (default) namespace, `'hash'` is the calculated hash, and `0` is the increment meaning how many times this exact hash has been found.

The resulting class name from this setup would be `'sc-hash'`, since zero increment is left out. If there was a collision, the second class name would be `'sc-hash-1'`.

Reply:
Maybe a bit confusing, index starts from zero so one means there's a collision.